working/0546-patterns/records-feature-specification.md

```diff
@@ -283,6 +283,41 @@ fields are) and collection literals.
 
 **TODO: Specify this more precisely.**
 
+### Constants
+
+A record expression is a compile-time constant expression
+if all its record field expressions are compile-time constant expressions. 
+
+_This is true whether the expression occurs in a constant context or not,
+which means that a record expression can be used directly as a parameter default value 
+if its record field expressions are constant expressions.
+Example: `f({(int, int) x = (1, 2)}) => ...`._
+
+A record expression is a potentially constant expression 
+if all its record field expressions are potentially constant or constant expressions.
+
+_This means that a record expression can be used in the initializer list
+of a constant non-redirecting generative constructor._
+
+Constant *object* instantiations create deeply immutable and canonicalied objects.
+Records are always unmodifiable, and if their field values are deeply immutable,
+like constants values, the records are also deeply immutable.
+It's meaningless to consider whether record constants are canonicalized,
+since records do not have a persistent identity.
+
+Because of that, there is no need for a `const (1, 2)` syntax to force a record 
+to be a constant, like there is for object creation expressions. 
```

Comment:
This is a really good insight! I like the idea of records being automagically const. They're immutable and have no persistent identity, so there's literally nothing preventing a compiler from treating any record expression with compile-time-known field values as a constant. And if that's the case, I see no compelling reason to require users to write `const` before it.

On the other hand, auto-const for records might lead users to want other immutable classes to be auto-constable too, and then get annoyed that they aren't. Perhaps something to think about with structs.

Reply:
True. 
My perspective is that where `const Foo(..)` creates a new `Foo` object, `(a, b)` does not create anything. It just places `a` next to `b`. It's just *two values* instead of *one value*.
It's not that you create a constant pair, it's that a pair of constant values are ... well, two constant values. They can be used wherever two constant values are needed.
You never need "a constant pair", you need a pair of constant values.


We can allow an initializer list of the form:
```dart
  const C((int, int) pair) : (this.x, this.y) = pair;
```
(We should!)

We could even allow 
```dart
  (final x, const y) = (1, 2);
```
(We probably shouldn't.)

Reply:
I don't think there is anything particularly tricky about

```dart
class C {
  final int x, y;
  const C((int, int) pair) : (x, y) = pair;
}
```

All we need is the ability to desugar the initializer list to `x = pair.$0, y = pair.$1`. But the opposite direction is more tricky:

```dart
class D {
  final (int, int) pair;
  const D(int x, int y) : pair = (x, y);
}
```

And even more:

```dart
class E {
  final Object o;
  const E(int x, int y, bool b) : o = b ? (x, y) : x + y;
}
```

Canonicalization is currently based on a compile-time version of `identical` applied to the values of each instance variable, but `identical` applied to a pair of records isn't guaranteed to _ever_ return true, so we wouldn't have any guarantees about canonicalization of instances of `E`, unless we specify that this compile-time-`identical` behaves differently.

Perhaps it is enough to say that `identical` at compile-time will _always_ perform a complete traversal in case it is invoked on two record values (of course, bailing out early if there is a non-match, but traversing the entire structure in order to obtain the value `true`).